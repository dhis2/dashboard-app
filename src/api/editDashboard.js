import { apiFetchDashboard } from './dashboards'

const onError = error => console.log('Error: ', error)

export const createDashboardMutation = {
    resource: 'dashboards',
    type: 'create',
    data: ({ data }) => data,
}

export const updateDashboardMutation = {
    resource: 'dashboards',
    type: 'update',
    id: ({ id }) => id,
    data: ({ data }) => data,
}

const generatePayload = (dashboard = {}, data) => {
    return {
        ...dashboard,
        name: data.name,
        description: data.description,
        dashboardItems: data.dashboardItems.map(item => ({
            ...item,
            width: item.w,
            height: item.h,
        })),
    }
}

export const updateDashboard = async (dataEngine, data) => {
    try {
        const dashboard = await apiFetchDashboard(dataEngine, data.id)

        const { response } = await dataEngine.mutate(updateDashboardMutation, {
            variables: {
                id: data.id,
                data: generatePayload(dashboard, data),
            },
        })

        return response.uid
    } catch (error) {
        onError(error)
    }
}

export const postDashboard = async (dataEngine, data) => {
    try {
        const { response } = await dataEngine.mutate(createDashboardMutation, {
            variables: {
                data: generatePayload({}, data),
            },
        })

        return response.uid
    } catch (error) {
        onError(error)
    }
}
