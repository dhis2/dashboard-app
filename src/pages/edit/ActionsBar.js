import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import i18n from '@dhis2/d2-i18n'
import TranslationDialog from '@dhis2/d2-ui-translation-dialog'
import { Button, ButtonStrip, FlyoutMenu, MenuItem } from '@dhis2/ui'
import { useDataEngine, useAlert } from '@dhis2/app-runtime'
import { useD2 } from '@dhis2/app-runtime-adapter-d2'

import FilterSettingsDialog from './FilterSettingsDialog'
import ConfirmActionDialog, {
    ACTION_DELETE,
    ACTION_DISCARD,
} from './ConfirmActionDialog'
import {
    tSaveDashboard,
    acClearEditDashboard,
    acSetPrintPreviewView,
    acClearPrintPreviewView,
    acSetFilterSettings,
    acUpdateDashboardLayout,
    acSetHideGrid,
} from '../../actions/editDashboard'
import { acClearPrintDashboard } from '../../actions/printDashboard'
import { tFetchDashboards } from '../../actions/dashboards'
import { acClearSelected } from '../../actions/selected'
import { deleteDashboardMutation } from './deleteDashboardMutation'
import {
    sGetEditDashboardRoot,
    sGetIsPrintPreviewView,
    sGetEditIsDirty,
    sGetEditDashboardItems,
} from '../../reducers/editDashboard'

import classes from './styles/ActionsBar.module.css'
import { getAutoItemShapes } from '../../modules/gridUtil'
import DropdownButton from '../../components/DropdownButton/DropdownButton'

const saveFailedMessage = i18n.t(
    'Failed to save dashboard. You might be offline or not have access to edit this dashboard.'
)

const deleteFailedMessage = i18n.t(
    'Failed to delete dashboard. You might be offline or not have access to edit this dashboard.'
)

const EditBar = ({ dashboard, ...props }) => {
    const { d2 } = useD2()
    const dataEngine = useDataEngine()
    const [translationDlgIsOpen, setTranslationDlgIsOpen] = useState(false)
    const [filterSettingsDlgIsOpen, setFilterSettingsDlgIsOpen] = useState(
        false
    )
    const [autoLayoutIsOpen, setAutoLayoutIsOpen] = useState(false)
    const [confirmDeleteDlgIsOpen, setConfirmDeleteDlgIsOpen] = useState(false)
    const [confirmDiscardDlgIsOpen, setConfirmDiscardDlgIsOpen] = useState(
        false
    )

    const [redirectUrl, setRedirectUrl] = useState(undefined)

    const saveFailureAlert = useAlert(saveFailedMessage, {
        critical: true,
    })

    const deleteFailureAlert = useAlert(deleteFailedMessage, {
        critical: true,
    })

    const onConfirmDelete = () => {
        setConfirmDeleteDlgIsOpen(true)
    }

    const onDeleteConfirmed = () => {
        setConfirmDeleteDlgIsOpen(false)

        dataEngine
            .mutate(deleteDashboardMutation, {
                variables: { id: dashboard.id },
            })
            .then(() => {
                props.clearSelected()

                return props.fetchDashboards()
            })
            .then(() => setRedirectUrl('/'))
            .catch(deleteFailureAlert.show)
    }

    const onSave = () => {
        props
            .saveDashboard()
            .then(newId => {
                props.clearSelected()
                setRedirectUrl(`/${newId}`)
            })
            .catch(() => saveFailureAlert.show())
    }

    const onPrintPreview = () => {
        if (props.isPrintPreviewView) {
            props.clearPrintPreview()
            props.clearPrintDashboard()
        } else {
            props.showPrintPreview()
        }
    }

    const onConfirmDiscard = () => {
        if (props.isDirty) {
            setConfirmDiscardDlgIsOpen(true)
        } else {
            onDiscardConfirmed()
        }
    }

    const onDiscardConfirmed = () => {
        props.onDiscardChanges()
        const redirectUrl = dashboard.id ? `/${dashboard.id}` : '/'

        setRedirectUrl(redirectUrl)
    }

    const onContinueEditing = () => {
        setConfirmDeleteDlgIsOpen(false)
        setConfirmDiscardDlgIsOpen(false)
    }

    const onFilterSettingsConfirmed = (
        filterSettingsRestrictability,
        selectedFilters
    ) => {
        const allowedFilters = filterSettingsRestrictability
            ? selectedFilters
            : []
        props.setFilterSettings({
            allowedFilters,
            restrictFilters: filterSettingsRestrictability,
        })
        toggleFilterSettingsDialog()
    }

    const toggleTranslationDialog = () =>
        setTranslationDlgIsOpen(!translationDlgIsOpen)

    const toggleFilterSettingsDialog = () => {
        setFilterSettingsDlgIsOpen(!filterSettingsDlgIsOpen)
    }

    const translationDialog = () =>
        dashboard.id ? (
            <TranslationDialog
                className="translation-dialog"
                d2={d2}
                open={translationDlgIsOpen}
                onRequestClose={toggleTranslationDialog}
                objectToTranslate={{
                    ...dashboard,
                    modelDefinition: { name: 'dashboard' },
                }}
                fieldsToTranslate={['name', 'description']}
                onTranslationError={err =>
                    console.log('translation update error', err)
                }
                onTranslationSaved={Function.prototype}
                insertTheme={true}
            />
        ) : null

    const filterSettingsDialog = () => {
        return (
            <FilterSettingsDialog
                restrictFilters={dashboard.restrictFilters}
                initiallySelectedItems={dashboard.allowedFilters}
                onClose={toggleFilterSettingsDialog}
                onConfirm={onFilterSettingsConfirmed}
                open={filterSettingsDlgIsOpen}
            />
        )
    }

    const getAutoLayoutMenuItem = numberOfCols => (
        <MenuItem
            dense
            label={`${numberOfCols} columns`}
            onClick={() => {
                setAutoLayoutIsOpen(false)
                Array.from(
                    document.querySelectorAll('.dashboard-scroll-container')
                ).forEach(el => (el.scrollTop = 0))
                props.onAutoLayoutSelect(numberOfCols)
            }}
        />
    )

    const getAutoLayoutMenu = () => (
        <FlyoutMenu>
            {new Array(10).fill().map((_, i) => getAutoLayoutMenuItem(i + 1))}
        </FlyoutMenu>
    )

    const renderActionButtons = () => (
        <ButtonStrip>
            <Button primary onClick={onSave} dataTest="save-dashboard-button">
                {i18n.t('Save changes')}
            </Button>
            <Button onClick={onPrintPreview}>
                {props.isPrintPreviewView
                    ? i18n.t('Exit Print preview')
                    : i18n.t('Print preview')}
            </Button>
            <Button onClick={toggleFilterSettingsDialog}>
                {i18n.t('Filter settings')}
            </Button>
            {dashboard.id && (
                <Button onClick={toggleTranslationDialog}>
                    {i18n.t('Translate')}
                </Button>
            )}
            {dashboard.id && dashboard.access?.delete && (
                <Button
                    onClick={onConfirmDelete}
                    dataTest="delete-dashboard-button"
                >
                    {i18n.t('Delete')}
                </Button>
            )}
            <DropdownButton
                onClick={() => {
                    setAutoLayoutIsOpen(!autoLayoutIsOpen)
                }}
                component={getAutoLayoutMenu()}
                // icon={<IconMore24 color={colors.grey700} />}
                open={autoLayoutIsOpen}
                // dataTest="delete-dashboard-button"
            >
                {i18n.t('Auto layout')}
            </DropdownButton>
        </ButtonStrip>
    )

    if (redirectUrl) {
        return <Redirect to={redirectUrl} />
    }

    const discardBtnText = dashboard.access?.update
        ? i18n.t('Exit without saving')
        : i18n.t('Go to dashboards')

    return (
        <>
            <div className={classes.editBar} data-test="edit-control-bar">
                <div className={classes.controls}>
                    {dashboard.access?.update ? renderActionButtons() : null}
                    <Button secondary onClick={onConfirmDiscard}>
                        {discardBtnText}
                    </Button>
                </div>
            </div>
            {dashboard.access?.update && filterSettingsDialog()}
            {dashboard.id && dashboard.access?.update && translationDialog()}
            {dashboard.id && dashboard.access?.delete && (
                <ConfirmActionDialog
                    action={ACTION_DELETE}
                    dashboardName={dashboard.name}
                    onConfirm={onDeleteConfirmed}
                    onCancel={onContinueEditing}
                    open={confirmDeleteDlgIsOpen}
                />
            )}
            <ConfirmActionDialog
                action={ACTION_DISCARD}
                onConfirm={onDiscardConfirmed}
                onCancel={onContinueEditing}
                open={confirmDiscardDlgIsOpen}
            />
        </>
    )
}

EditBar.propTypes = {
    clearPrintDashboard: PropTypes.func,
    clearPrintPreview: PropTypes.func,
    clearSelected: PropTypes.func,
    dashboard: PropTypes.object,
    fetchDashboards: PropTypes.func,
    isDirty: PropTypes.bool,
    isPrintPreviewView: PropTypes.bool,
    saveDashboard: PropTypes.func,
    setFilterSettings: PropTypes.func,
    showPrintPreview: PropTypes.func,
    onAutoLayoutSelect: PropTypes.func,
    onDiscardChanges: PropTypes.func,
}

const mapStateToProps = state => {
    return {
        dashboard: sGetEditDashboardRoot(state),
        isPrintPreviewView: sGetIsPrintPreviewView(state),
        isDirty: sGetEditIsDirty(state),
    }
}

const mapDispatchToProps = {
    clearPrintDashboard: () => dispatch => dispatch(acClearPrintDashboard()),
    clearPrintPreview: () => dispatch => dispatch(acClearPrintPreviewView()),
    clearSelected: () => dispatch => dispatch(acClearSelected()),
    saveDashboard: () => dispatch => dispatch(tSaveDashboard()).then(id => id),
    fetchDashboards: () => dispatch => dispatch(tFetchDashboards()),
    onDiscardChanges: () => dispatch => dispatch(acClearEditDashboard()),
    setFilterSettings: value => dispatch =>
        dispatch(acSetFilterSettings(value)),
    showPrintPreview: () => dispatch => dispatch(acSetPrintPreviewView()),
    onAutoLayoutSelect: value => (dispatch, getState) => {
        const prevItems = sGetEditDashboardItems(getState())
        const itemsWithNewShapes = getAutoItemShapes(prevItems, value)
        dispatch(acSetHideGrid(true))
        dispatch(acUpdateDashboardLayout(itemsWithNewShapes))
        setTimeout(() => dispatch(acSetHideGrid(false)), 0)
    },
}

export default connect(mapStateToProps, mapDispatchToProps)(EditBar)
