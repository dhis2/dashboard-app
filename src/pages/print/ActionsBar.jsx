import i18n from '@dhis2/d2-i18n'
import { Button, colors, IconChevronLeft24 } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React from 'react'
import { Link } from 'react-router-dom'
import classes from './styles/PrintActionsBar.module.css'

// set in PrintActionsBar.module.css
export const PRINT_ACTIONS_BAR_HEIGHT = 44
export const PRINT_ACTIONS_BAR_HEIGHT_SM = 36

const PrintActionsBar = ({ id }) => {
    const getExitPrintButton = (isSmall) => (
        <Button
            className={isSmall ? classes.buttonSmall : classes.buttonLarge}
            small={isSmall}
        >
            <IconChevronLeft24 color={colors.grey700} />
            {i18n.t('Exit print preview')}
        </Button>
    )

    const getPrintButton = (isSmall) => (
        <Button
            className={isSmall ? classes.buttonSmall : classes.buttonLarge}
            small={isSmall}
            onClick={window.print}
        >
            {i18n.t('Print')}
        </Button>
    )

    return (
        <div className={classes.container}>
            <div className={classes.actions}>
                <Link className={classes.link} to={`/${id}`}>
                    {getExitPrintButton(true)}
                    {getExitPrintButton(false)}
                </Link>
                {getPrintButton(true)}
                {getPrintButton(false)}
            </div>
        </div>
    )
}

PrintActionsBar.defaultValues = {
    id: '/',
}

PrintActionsBar.propTypes = {
    id: PropTypes.string,
}

export default PrintActionsBar
