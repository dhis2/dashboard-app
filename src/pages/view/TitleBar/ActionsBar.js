import { useDataEngine, useAlert, useOnlineStatus } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import {
    Button,
    FlyoutMenu,
    colors,
    IconMore24,
    SharingDialog,
} from '@dhis2/ui'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { connect } from 'react-redux'
import { Link, Redirect } from 'react-router-dom'
import { acSetDashboardStarred } from '../../../actions/dashboards.js'
import { acClearItemFilters } from '../../../actions/itemFilters.js'
import { acSetShowDescription } from '../../../actions/showDescription.js'
import { apiPostShowDescription } from '../../../api/description.js'
import ConfirmActionDialog from '../../../components/ConfirmActionDialog.js'
import DropdownButton from '../../../components/DropdownButton/DropdownButton.js'
import MenuItem from '../../../components/MenuItemWithTooltip.js'
import OfflineTooltip from '../../../components/OfflineTooltip.js'
import { useCacheableSection } from '../../../modules/useCacheableSection.js'
import { orObject } from '../../../modules/util.js'
import { sGetDashboardStarred } from '../../../reducers/dashboards.js'
import { sGetNamedItemFilters } from '../../../reducers/itemFilters.js'
import { sGetSelected } from '../../../reducers/selected.js'
import { sGetShowDescription } from '../../../reducers/showDescription.js'
import { ROUTE_START_PATH } from '../../start/index.js'
import { apiStarDashboard } from './apiStarDashboard.js'
import FilterSelector from './FilterSelector.js'
import StarDashboardButton from './StarDashboardButton.js'
import classes from './styles/ActionsBar.module.css'

const ViewActions = ({
    id,
    access,
    showDescription,
    starred,
    setDashboardStarred,
    updateShowDescription,
    removeAllFilters,
    restrictFilters,
    allowedFilters,
    filtersLength,
}) => {
    const [moreOptionsSmallIsOpen, setMoreOptionsSmallIsOpen] = useState(false)
    const [moreOptionsIsOpen, setMoreOptionsIsOpen] = useState(false)
    const [sharingDialogIsOpen, setSharingDialogIsOpen] = useState(false)
    const [confirmCacheDialogIsOpen, setConfirmCacheDialogIsOpen] =
        useState(false)
    const [redirectUrl, setRedirectUrl] = useState(null)
    const dataEngine = useDataEngine()
    const { offline } = useOnlineStatus()
    const { lastUpdated, isCached, startRecording, remove } =
        useCacheableSection(id)

    const { show } = useAlert(
        ({ msg }) => msg,
        ({ isCritical }) =>
            isCritical ? { critical: true } : { warning: true }
    )

    const toggleMoreOptions = (small) =>
        small
            ? setMoreOptionsSmallIsOpen(!moreOptionsSmallIsOpen)
            : setMoreOptionsIsOpen(!moreOptionsIsOpen)

    const closeMoreOptions = () => {
        setMoreOptionsSmallIsOpen(false)
        setMoreOptionsIsOpen(false)
    }

    if (redirectUrl) {
        return <Redirect to={redirectUrl} />
    }

    const onRecordError = () => {
        show({
            msg: i18n.t(
                "The dashboard couldn't be made available offline. Try again."
            ),
            isCritical: true,
        })
    }

    const onCacheDashboardConfirmed = () => {
        setConfirmCacheDialogIsOpen(false)
        removeAllFilters()
        startRecording({
            onError: onRecordError,
        })
    }

    const onRemoveFromOffline = () => {
        closeMoreOptions()
        lastUpdated && remove()
    }

    const onAddToOffline = () => {
        closeMoreOptions()
        return filtersLength
            ? setConfirmCacheDialogIsOpen(true)
            : startRecording({
                  onError: onRecordError,
              })
    }

    const onToggleShowDescription = () => {
        updateShowDescription(!showDescription)
        closeMoreOptions()
        !offline && apiPostShowDescription(!showDescription)
    }

    const onToggleStarredDashboard = () =>
        apiStarDashboard(dataEngine, id, !starred)
            .then(() => {
                setDashboardStarred(id, !starred)
                closeMoreOptions()
            })
            .catch(() => {
                const msg = starred
                    ? i18n.t('Failed to unstar the dashboard')
                    : i18n.t('Failed to star the dashboard')
                show({ msg, isCritical: false })
            })

    const onToggleSharingDialog = () =>
        setSharingDialogIsOpen(!sharingDialogIsOpen)

    const userAccess = orObject(access)

    const getMoreMenu = () => (
        <FlyoutMenu>
            {lastUpdated ? (
                <MenuItem
                    dense
                    disabledWhenOffline={false}
                    label={i18n.t('Remove from offline storage')}
                    onClick={onRemoveFromOffline}
                />
            ) : (
                <MenuItem
                    dense
                    disabled={offline}
                    label={i18n.t('Make available offline')}
                    onClick={onAddToOffline}
                />
            )}
            {lastUpdated && (
                <MenuItem
                    dense
                    label={i18n.t('Sync offline data now')}
                    disabled={offline}
                    onClick={onAddToOffline}
                />
            )}
            <MenuItem
                dense
                disabled={offline}
                label={
                    starred
                        ? i18n.t('Unstar dashboard')
                        : i18n.t('Star dashboard')
                }
                onClick={onToggleStarredDashboard}
            />
            <MenuItem
                dense
                disabledWhenOffline={false}
                label={
                    showDescription
                        ? i18n.t('Hide description')
                        : i18n.t('Show description')
                }
                onClick={onToggleShowDescription}
            />
            <MenuItem
                dense
                disabled={offline && !isCached}
                disabledWhenOffline={false}
                label={i18n.t('Print')}
                dataTest="print-menu-item"
            >
                <MenuItem
                    dense
                    disabled={offline && !isCached}
                    disabledWhenOffline={false}
                    label={i18n.t('Dashboard layout')}
                    onClick={() => setRedirectUrl(`${id}/printlayout`)}
                    dataTest="print-layout-menu-item"
                />
                <MenuItem
                    dense
                    disabled={offline && !isCached}
                    disabledWhenOffline={false}
                    label={i18n.t('One item per page')}
                    onClick={() => setRedirectUrl(`${id}/printoipp`)}
                    dataTest="print-oipp-menu-item"
                />
            </MenuItem>
            <Link to={ROUTE_START_PATH} className={classes.link}>
                <MenuItem
                    dense
                    disabledWhenOffline={false}
                    label={i18n.t('Close dashboard')}
                />
            </Link>
        </FlyoutMenu>
    )

    const getMoreButton = (className, useSmall) => (
        <DropdownButton
            className={className}
            small={useSmall}
            open={useSmall ? moreOptionsSmallIsOpen : moreOptionsIsOpen}
            disabledWhenOffline={false}
            onClick={() => toggleMoreOptions(useSmall)}
            icon={<IconMore24 color={colors.grey700} />}
            component={getMoreMenu()}
        >
            {i18n.t('More')}
        </DropdownButton>
    )

    return (
        <>
            <div className={classes.actions}>
                <StarDashboardButton
                    starred={starred}
                    onClick={onToggleStarredDashboard}
                />
                <div className={classes.strip}>
                    {userAccess.update ? (
                        <OfflineTooltip>
                            <Button
                                disabled={offline}
                                className={classes.editButton}
                                onClick={() => setRedirectUrl(`${id}/edit`)}
                            >
                                {i18n.t('Edit')}
                            </Button>
                        </OfflineTooltip>
                    ) : null}
                    {userAccess.manage ? (
                        <OfflineTooltip>
                            <Button
                                disabled={offline}
                                className={classes.shareButton}
                                onClick={onToggleSharingDialog}
                            >
                                {i18n.t('Share')}
                            </Button>
                        </OfflineTooltip>
                    ) : null}
                    <FilterSelector
                        allowedFilters={allowedFilters}
                        restrictFilters={restrictFilters}
                    />
                    {getMoreButton(classes.moreButton, false)}
                    {getMoreButton(classes.moreButtonSmall, true)}
                </div>
            </div>
            {id && sharingDialogIsOpen && (
                <SharingDialog
                    id={id}
                    type="dashboard"
                    onClose={onToggleSharingDialog}
                />
            )}
            <ConfirmActionDialog
                title={i18n.t('Clear dashboard filters?')}
                message={i18n.t(
                    "A dashboard's filters canâ€™t be saved offline. Do you want to remove the filters and make this dashboard available offline?"
                )}
                cancelLabel={i18n.t('No, cancel')}
                confirmLabel={i18n.t('Yes, clear filters and sync')}
                onConfirm={onCacheDashboardConfirmed}
                onCancel={() => setConfirmCacheDialogIsOpen(false)}
                open={confirmCacheDialogIsOpen}
            />
        </>
    )
}

ViewActions.propTypes = {
    access: PropTypes.object,
    allowedFilters: PropTypes.array,
    filtersLength: PropTypes.number,
    id: PropTypes.string,
    removeAllFilters: PropTypes.func,
    restrictFilters: PropTypes.bool,
    setDashboardStarred: PropTypes.func,
    showDescription: PropTypes.bool,
    starred: PropTypes.bool,
    updateShowDescription: PropTypes.func,
}

const mapStateToProps = (state) => {
    const dashboard = sGetSelected(state)

    return {
        ...dashboard,
        filtersLength: sGetNamedItemFilters(state).length,
        starred: dashboard.id
            ? sGetDashboardStarred(state, dashboard.id)
            : false,
        showDescription: sGetShowDescription(state),
    }
}

export default connect(mapStateToProps, {
    setDashboardStarred: acSetDashboardStarred,
    removeAllFilters: acClearItemFilters,
    updateShowDescription: acSetShowDescription,
})(ViewActions)
