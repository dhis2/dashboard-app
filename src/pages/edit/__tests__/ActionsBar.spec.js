import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import { DEFAULT_STATE_EDIT_DASHBOARD } from '../../../reducers/editDashboard.js'
import ActionsBar from '../ActionsBar.js'

const mockStore = configureMockStore()

jest.mock('@dhis2/app-runtime')

/* eslint-disable react/prop-types */
jest.mock('@dhis2/ui', () => {
    const originalModule = jest.requireActual('@dhis2/ui')

    return {
        __esModule: true,
        ...originalModule,
        ButtonStrip: function Mock({ children }) {
            return <div className="ui-ButtonStrip">{children}</div>
        },
        Button: function Mock({ children }) {
            return <div className="ui-Button">{children}</div>
        },
    }
})
/* eslint-enable react/prop-types */

/* eslint-disable react/prop-types */
jest.mock('@dhis2/analytics', () => {
    const originalModule = jest.requireActual('@dhis2/analytics')

    return {
        __esModule: true,
        ...originalModule,
        OfflineTooltip: function Mock({ children }) {
            return <div className="OfflineTooltip">{children}</div>
        },
        useCachedDataQuery: () => ({
            currentUser: {
                username: 'rainbowDash',
                id: 'r3nb0d5h',
            },
        }),
    }
})
/* eslint-enable react/prop-types */

jest.mock(
    '../FilterSettingsDialog',
    () =>
        function Mock() {
            return <div className="mock-filter-settings-dialog" />
        }
)

/* eslint-disable react/prop-types */
jest.mock(
    '../../../components/ConfirmActionDialog',
    () =>
        function MockConfirmActionDialog({ open }) {
            return open ? <div className="mock-confirm-action-dialog" /> : null
        }
)
/* eslint-enable react/prop-types */

jest.mock('@dhis2/app-runtime', () => ({
    useDhis2ConnectionStatus: jest.fn(() => ({
        isConnected: true,
        isDisconnected: false,
    })),
    useDataEngine: jest.fn(() => ({ dataEngine: {} })),
    useAlert: jest.fn(() => ({})),
}))

test('renders the ActionsBar without Delete when no delete access', async () => {
    const store = {
        editDashboard: {
            id: 'rainbowDash',
            name: 'Rainbow Dash',
            access: {
                update: true,
                delete: false,
            },
            printPreviewView: false,
        },
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <ActionsBar />
        </Provider>
    )

    expect(container).toMatchSnapshot()
})

test('renders only the Go to Dashboards button when no update access', async () => {
    const store = {
        editDashboard: {
            id: 'rainbowDash',
            name: 'Rainbow Dash',
            access: {
                update: false,
                delete: false,
            },
            printPreviewView: false,
        },
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <ActionsBar />
        </Provider>
    )

    expect(container).toMatchSnapshot()
})

test('renders Save and Discard buttons but not translation dialog when new dashboard (no dashboard id)', async () => {
    const store = {
        editDashboard: DEFAULT_STATE_EDIT_DASHBOARD,
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <ActionsBar />
        </Provider>
    )

    expect(container).toMatchSnapshot()
})

test('renders Translate, Delete, and Discard buttons when delete access', async () => {
    const store = {
        editDashboard: {
            id: 'rainbowDash',
            name: 'Rainbow Dash',
            access: {
                update: true,
                delete: true,
            },
            printPreviewView: false,
        },
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <ActionsBar />
        </Provider>
    )

    expect(container).toMatchSnapshot()
})
