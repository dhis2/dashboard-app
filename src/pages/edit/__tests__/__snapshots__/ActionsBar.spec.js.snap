// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Save and Discard buttons but not translation dialog when new dashboard (no dashboard id) 1`] = `
<div>
  <div
    class="editBar"
    data-test="edit-control-bar"
  >
    <div
      class="controls"
    >
      <div
        class="jsx-71743532 start"
        data-test="dhis2-uicore-buttonstrip"
      >
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 primary"
            data-test="save-dashboard-button"
            type="button"
          >
            Save changes
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="dhis2-uicore-button"
            type="button"
          >
            Print preview
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="dhis2-uicore-button"
            type="button"
          >
            Filter settings
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        />
        <div
          class="jsx-71743532 box"
        />
        <div
          class="jsx-71743532 box"
        >
          <div>
            <button
              class="jsx-2371629422 "
              data-test="dhis2-uicore-button"
              type="button"
            >
              Auto layout
              <svg
                class="arrow"
                viewBox="0 0 12 12"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m5.29289 8.7071c.39053.3905 1.02369.3905 1.41422 0l2.99999-2.99999c.3905-.39053.3905-1.02369 0-1.41422-.3905-.39052-1.0237-.39052-1.4142 0l-2.2929 2.2929-2.29289-2.2929c-.39053-.39052-1.02369-.39052-1.41422 0-.39052.39053-.39052 1.02369 0 1.41422z"
                />
                <style>
                  
                svg {
                    fill: inherit;
                    height: 12px;
                    width: 12px;
                    vertical-align: middle;
                    pointer-events: none;
                }
            
                </style>
              </svg>
            </button>
          </div>
        </div>
      </div>
      <button
        class="jsx-2371629422 secondary"
        data-test="dhis2-uicore-button"
        type="button"
      >
        Exit without saving
      </button>
    </div>
  </div>
  <div
    class="mock-filter-settings-dialog"
  />
  
  
</div>
`;

exports[`renders Translate, Delete, and Discard buttons when delete access 1`] = `
<div>
  <div
    class="editBar"
    data-test="edit-control-bar"
  >
    <div
      class="controls"
    >
      <div
        class="jsx-71743532 start"
        data-test="dhis2-uicore-buttonstrip"
      >
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 primary"
            data-test="save-dashboard-button"
            type="button"
          >
            Save changes
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="dhis2-uicore-button"
            type="button"
          >
            Print preview
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="dhis2-uicore-button"
            type="button"
          >
            Filter settings
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="dhis2-uicore-button"
            type="button"
          >
            Translate
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="delete-dashboard-button"
            type="button"
          >
            Delete
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <div>
            <button
              class="jsx-2371629422 "
              data-test="dhis2-uicore-button"
              type="button"
            >
              Auto layout
              <svg
                class="arrow"
                viewBox="0 0 12 12"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m5.29289 8.7071c.39053.3905 1.02369.3905 1.41422 0l2.99999-2.99999c.3905-.39053.3905-1.02369 0-1.41422-.3905-.39052-1.0237-.39052-1.4142 0l-2.2929 2.2929-2.29289-2.2929c-.39053-.39052-1.02369-.39052-1.41422 0-.39052.39053-.39052 1.02369 0 1.41422z"
                />
                <style>
                  
                svg {
                    fill: inherit;
                    height: 12px;
                    width: 12px;
                    vertical-align: middle;
                    pointer-events: none;
                }
            
                </style>
              </svg>
            </button>
          </div>
        </div>
      </div>
      <button
        class="jsx-2371629422 secondary"
        data-test="dhis2-uicore-button"
        type="button"
      >
        Exit without saving
      </button>
    </div>
  </div>
  <div
    class="mock-filter-settings-dialog"
  />
  <div
    class="mock-translation-dialog"
  />
</div>
`;

exports[`renders only the Go to Dashboards button when no update access 1`] = `
<div>
  <div
    class="editBar"
    data-test="edit-control-bar"
  >
    <div
      class="controls"
    >
      <button
        class="jsx-2371629422 secondary"
        data-test="dhis2-uicore-button"
        type="button"
      >
        Go to dashboards
      </button>
    </div>
  </div>
</div>
`;

exports[`renders the ActionsBar without Delete when no delete access 1`] = `
<div>
  <div
    class="editBar"
    data-test="edit-control-bar"
  >
    <div
      class="controls"
    >
      <div
        class="jsx-71743532 start"
        data-test="dhis2-uicore-buttonstrip"
      >
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 primary"
            data-test="save-dashboard-button"
            type="button"
          >
            Save changes
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="dhis2-uicore-button"
            type="button"
          >
            Print preview
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="dhis2-uicore-button"
            type="button"
          >
            Filter settings
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 "
            data-test="dhis2-uicore-button"
            type="button"
          >
            Translate
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        />
        <div
          class="jsx-71743532 box"
        >
          <div>
            <button
              class="jsx-2371629422 "
              data-test="dhis2-uicore-button"
              type="button"
            >
              Auto layout
              <svg
                class="arrow"
                viewBox="0 0 12 12"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m5.29289 8.7071c.39053.3905 1.02369.3905 1.41422 0l2.99999-2.99999c.3905-.39053.3905-1.02369 0-1.41422-.3905-.39052-1.0237-.39052-1.4142 0l-2.2929 2.2929-2.29289-2.2929c-.39053-.39052-1.02369-.39052-1.41422 0-.39052.39053-.39052 1.02369 0 1.41422z"
                />
                <style>
                  
                svg {
                    fill: inherit;
                    height: 12px;
                    width: 12px;
                    vertical-align: middle;
                    pointer-events: none;
                }
            
                </style>
              </svg>
            </button>
          </div>
        </div>
      </div>
      <button
        class="jsx-2371629422 secondary"
        data-test="dhis2-uicore-button"
        type="button"
      >
        Exit without saving
      </button>
    </div>
  </div>
  <div
    class="mock-filter-settings-dialog"
  />
  <div
    class="mock-translation-dialog"
  />
</div>
`;
